### ISSUE TITLE
feature request: run tests in parallel

### ISSUE THREAD

**Issue by wood1986** [jasmine/jasmine-npm#152]
as I am trying to add the parallel feature to jasmine, this is my [POC](https://github.com/wood1986/jasmine-npm/commit/9a0cc4821ff1645ce887ca0e58b118f45d4ea3bd)

But the downside of this implementation is I keep creating the worker. The reason of the creation is I cannot reuse Jasmine object after finishing the spec. Is there a way to reuse the Jasmine object or call `jasmine.execute([file], env.filter)` after the completion?

**Comment by slackersoft** [comment]
A given Jasmine `Env` (from jasmine-core) does not expect to be `execute`d more than once, and we don't currently have a mechanism to do something like reuse the context and configuration, but not the actual tree of all the suites and specs to be executed. I'm not entirely sure there's much benefit to doing the reuse, given how Jasmine currently works under the hood, are you seeing performance (or other) issues with re-creating the Jasmine environment?

**Comment by wood1986** [comment]
If I keep creating the worker for each spec file, all the tests are passed but the execution is longer than a single process. Then I try to create Jasmine objects for each spec file. When the worker finishes the test for one spec file, it will create another Jasmine object for testing the next spec file. But some tests are failed.

I guess the Jasmine object mutated something in the global object. This makes the test failed.

![image](https://user-images.githubusercontent.com/5212215/54586113-496b2400-49d9-11e9-9295-f3a025799404.png)

Here is my updated commit 
https://github.com/wood1986/jasmine-npm/commit/bc5e76afa5cfba3b50347f235598639672f7fd31

**Comment by slackersoft** [comment]
Those functions are all set up in the `helpers` from the jasmine.json, so it looks like they might not be getting required correctly when you're spinning up the workers. There may also be other things that would make Jasmine a better citizen in a multi-use scenario (i.e. perhaps not installing itself onto `global`/`window` by default). Fixing some of these might also make it easier to add a parallel runner.

Additionally, since you're looking at this, I would like to see parallelization of a suite be optional (at least to start) and not the default.

**Comment by wood1986** [comment]
I know why. It is because of the node's `require` 

At the first run, the helpers have not been `require`. Then the initialization for the `jasmine.getEnv()` in each helper file will get executed. If I reuse the worker and recreate a new Jasmine object, the `require` for each helper file will not initialize the `jasmine.getEnv()` again.

I may add this `delete require.cache[require.resolve('fileName')];` before calling the `require`

**Comment by wood1986** [comment]
||Before|After|
|---|---|---|
|real|0m2.667s|0m2.286s|
|user|0.m2.015s|0m4.273s|
|sys|0m0.272s|0m0.727s|

Before
![before](https://user-images.githubusercontent.com/5212215/54736473-726def00-4b68-11e9-9a19-70d6f1608e30.gif)

After with 4 cores
![after](https://user-images.githubusercontent.com/5212215/54736502-8ade0980-4b68-11e9-8131-7c384ef0b989.gif)

Basically, it has a 0.5s improvement.

@slackersoft I can add this feature but there are several things I need to think about.
1. Reimplement a way to consolidate the results when it is a cluster mode
2. Random seed seems useless in parallel mode because [`jasmine.execute([one file])`](https://github.com/jasmine/jasmine-npm/blob/b5d9ef9381bcded54c27e63664506b7de0ca546d/lib/command.js#L158) will only execute one spec file in my implementation. The scope of the randomization will be in one single file.
3. My implementation of the parallel is completely different from the existing. I will have a queue in master. When the worker finishes the tests for a single file, it will notify and ask the master for the next spec file. The existing implementation is a big array of spec files pass to `jasmine.execute`. In short. when the concurrency is set to 1, the behavior is completely different from the existing. Nothing I can reuse and I may need to introduce another reporter for the concurrent situation.

Any thoughts or recommendation? Do you want to add this feature?

**Comment by wood1986** [comment]
See #153

**Comment by slackersoft** [comment]
Closing this. #153 has been merged.

**Comment by bodia-uz** [comment]
@slackersoft any chance the work parallel tests will continue?

**Comment by slackersoft** [comment]
Getting Jasmine-Core ready to be able to run in parallel would take a lot of work to get right, and that work would be necessary for Jasmine-Npm to really support parallelization. We'd be happy to review some PRs to make it possible to move in this direction.

**Comment by sgravrock** [comment]
Parallel support is now available in 5.0.0-alpha.0. See <https://jasmine.github.io/tutorials/running_specs_in_parallel> for details.

**Comment by ajvincent** [comment]
@sgravrock I've wanted for a very long to deeply apologize for my behavior on the other issue.  

I was _way, way_ too excited about this feature, and I lost my senses (particularly in the "playing well with others" category) in trying to rapidly put together a working model.  I didn't properly weigh "project owners" versus "users" (which is why I kept flip-flopping as suggestions / directives came in), and as you pointed out, I probably underestimated the complexity involved by a lot.

I absolutely did not mean to offend you, or to make you think I wasn't listening.  My screw-up was in listening to everyone, and going with "the last person to comment usually wins".  Which - as I'm now discovering from another GitHub repository, from the perspective of an owner - is really bad policy.

This has been a moment of great shame for me, personally.  Locking the thread emphasized for me very strongly just how badly I offended you.

I don't fault you one bit for this.  In fact, I'm grateful to you for finally implementing this feature, and I know you put in a ton of work to get to this point.  I only wish I had slowed down before I crashed so hard.

I am looking forward to this going to final release.  I am very sorry that I was so bull-headed when I attempted to implement this, and I would appreciate an opportunity to make amends in some way.

Alex
